### ISSUE TITLE
slice_input_producer bug (slows down exponentially for larger datasets)

### ISSUE THREAD

**Issue by jrbtaylor** [tensorflow/tensorflow#11591]
My input pipeline involves sampling a filename and label with slice_input_producer, then data augmentation, then batching with tf.train.shuffle_batch. For smaller datasets (<1M), this works fine. For larger datasets, slice_input_producer slows down, even though all it's doing is sampling from two lists. It's bad enough that it's several orders of magnitude slower than all the rest of my input pipeline and training combined. I wrote a quick test to measure the time per call to slice_input_producer for different lengths of input lists:
```
import time

import tensorflow as tf

rept = 1000
for datasize in [1e3,1e6,1e7,1e8]:
    sym_x = tf.train.slice_input_producer(
        [list(range(int(datasize)))], shuffle=True,
        capacity=10)
    with tf.Session() as sess:
        sess.run(tf.local_variables_initializer())
        sess.run(tf.global_variables_initializer())
        coord = tf.train.Coordinator()
        threads = tf.train.start_queue_runners(sess=sess, coord=coord)
        start_time = time.time()
        for r in range(rept):
            x = sess.run(sym_x)
        print('Datalength: %i  ---  time = %.3f ms/sample'
              % (datasize,(time.time()-start_time)/rept*1000))
        coord.request_stop()
        coord.join(threads)
        sess.close()
```

Which gives:

> Datalength: 1000  ---  time = 0.172 ms/sample
> Datalength: 1000000  ---  time = 0.207 ms/sample
> Datalength: 10000000  ---  time = 0.537 ms/sample
> Datalength: 100000000  ---  time = 13.991 ms/sample

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
('v1.2.0-0-g12f033d', '1.2.0')

**Comment by drpngx** [comment]
Thanks for reporting. If it's new code, I would recommend using `tf.contrib.data`.

/CC: @mrry

**Comment by jrbtaylor** [comment]
I wrote a similar test for the tf.contrib.data functions I would be using for the same task above:

```
import time

import tensorflow as tf
from tensorflow.python.framework import ops,dtypes
from tensorflow.contrib.data import Dataset
from tensorflow.contrib.data import Iterator

rept = 1000
for datasize in [1e3, 1e6, 1e7, 1e8]:
    data = list(range(int(datasize)))
    data_tensor = ops.convert_to_tensor(data)
    dataset = Dataset.from_tensor_slices(data_tensor)
    dataset = dataset.shuffle(int(datasize))
    iterator = Iterator.from_structure(dataset.output_types,
                                       dataset.output_shapes)
    init_op = iterator.make_initializer(dataset)
    next_element = iterator.get_next()

    with tf.Session() as sess:
        sess.run(init_op)
        start_time = time.time()
        for r in range(rept):
            x = sess.run(next_element)
        print('Datalength: %i  ---  time = %.3f ms/sample'%(
        datasize, (time.time()-start_time)/rept*1000))
        sess.close()
```

It actually does worse:

> Datalength: 1000  ---  time = 0.113 ms/sample
> Datalength: 1000000  ---  time = 0.374 ms/sample
> Datalength: 10000000  ---  time = 2.647 ms/sample
> Datalength: 100000000  ---  time = 27.247 ms/sample

In both cases (tf.contrib.data and the old functions above), the overhead appears to be in shuffling larger datasets. The tf.contrib.data functions at least allow access to the buffer size for shuffling, so I can do shuffling on subsets of the data (not true random, but better than nothing). This solves the problem in practice for large datasets. Why not offer an option to do shuffling by random sampling with replacement? This would scale better - there's no need to actually shuffle a huge list, just sample randomly.

In practice, tf.contrib.data is about 3x slower than my old input pipeline on real disk reads, so @drpngx maybe you could expand on why you recommend tf.contrib.data. Is there a reason beyond being able to set the buffer size for the shuffle operation?

**Comment by mrry** [comment]
Just to clarify, those two programs aren't equivalent. In the `tf.contrib.data` version, the equivalent to `capacity` is set to `int(datasize)`, whereas the `slice_input_producer` version uses `capacity=10`. This means that your `tf.contrib.data` version has to accumulate the entire input before producing the first record, whereas the `slice_input_producer` version accumulates only 10 records. The following program has more encouraging behavior:

```python
import time

import tensorflow as tf
from tensorflow.python.framework import ops,dtypes
from tensorflow.contrib.data import Dataset
from tensorflow.contrib.data import Iterator

rept = 1000
for datasize in [1e3, 1e6, 1e7, 1e8]:
  with tf.Graph().as_default():
    data = list(range(int(datasize)))
    data_tensor = ops.convert_to_tensor(data)
    dataset = Dataset.from_tensor_slices(data_tensor)
    #dataset = Dataset.range(int(datasize))
    dataset = dataset.shuffle(10)  #int(datasize))
    iterator = Iterator.from_structure(dataset.output_types,
                                       dataset.output_shapes)
    init_op = iterator.make_initializer(dataset)
    next_element = iterator.get_next()

    with tf.Session() as sess:
        sess.run(init_op)
        start_time = time.time()
        for r in range(rept):
            x = sess.run(next_element)
        print('Datalength: %i  ---  time = %.3f ms/sample'%(
        datasize, (time.time()-start_time)/rept*1000))
```

> Datalength: 1000  ---  time = 0.123 ms/sample
> Datalength: 1000000  ---  time = 0.109 ms/sample
> Datalength: 10000000  ---  time = 0.102 ms/sample
> Datalength: 100000000  ---  time = 0.714 ms/sample

I suspect the problem actually stems from this line: 

```python
    data_tensor = ops.convert_to_tensor(data)
```

Embedding large tensors in the graph is slow, and you pay a noticeable and increasing penalty for both the `ops.convert_to_tensor()` call and the first `sess.run()` call after converting the tensor. Essentially, the tensor is copied several times, serialized and deserialized. I think the `1e8` case is slower because my local machine was running out of memory due to all of the copies!

We haven't historically used random sampling with replacement in TensorFlow because it's tricky to scale it to datasets that are larger than memory. However, it would certainly be possible to add a `Dataset` that implemented it.

**Comment by jrbtaylor** [comment]
Maybe a better question - how should I be dealing with large datasets then? For instance, with 1e7 or 1e8 filenames, how do I sample from that if I don't put the filenames into a tensor?

**Comment by mrry** [comment]
That's a lot of filenames! Perhaps you could write them to an index file, one filename per line, and use `TextLineDataset(index_filename).flat_map(lambda filename: ...)` where `...` is the appropriate reading and parsing logic for a single file.

**Comment by reedwm** [comment]
Closing since a solution to your last question has been provided.

**Comment by jrbtaylor** [comment]
@mrry What do you mean by index file? What format should I be writing the filenames/labels to for `TextLineDataset`?

**Comment by kvanhoey** [comment]
Hello.

I experience the same behavior as @jrbtaylor on a much smaller dataset already (the MNIST data, 60K training examples).
Shuffling progressively slows down the learning iterations, up until the end of an epoch, then it continues at the initial speed again:

- No shuffling:

> Elapsed time for 100 iterations: 0.7071466445922852
Elapsed time for 100 iterations: 0.928595781326294
Elapsed time for 100 iterations: 0.9202249050140381
Elapsed time for 100 iterations: 0.9153792858123779
Elapsed time for 100 iterations: 0.9065940380096436
Elapsed time for 100 iterations: 0.9027049541473389
EPOCH++
Elapsed time for 100 iterations: 0.903191328048706
Elapsed time for 100 iterations: 0.9209485054016113
Elapsed time for 100 iterations: 0.9005069732666016
Elapsed time for 100 iterations: 0.9149377346038818
Elapsed time for 100 iterations: 0.9105048179626465
Elapsed time for 100 iterations: 0.8968446254730225
EPOCH++
Elapsed time for 100 iterations: 0.899205207824707
Elapsed time for 100 iterations: 0.9078717231750488
Elapsed time for 100 iterations: 0.9038975238800049
Elapsed time for 100 iterations: 0.9130148887634277
Elapsed time for 100 iterations: 0.9119799137115479
Elapsed time for 100 iterations: 0.9025707244873047

- With shuffling
> Elapsed time for 100 iterations: 0.756305456161499
Elapsed time for 100 iterations: 0.9620733261108398
Elapsed time for 100 iterations: 1.1773226261138916
Elapsed time for 100 iterations: 2.098752498626709
Elapsed time for 100 iterations: 3.113154411315918
Elapsed time for 100 iterations: 3.66085147857666
EPOCH++
Elapsed time for 100 iterations: 1.9388504028320312
Elapsed time for 100 iterations: 0.9970040321350098
Elapsed time for 100 iterations: 1.2153730392456055
Elapsed time for 100 iterations: 2.1299166679382324
Elapsed time for 100 iterations: 3.369262933731079
Elapsed time for 100 iterations: 3.6159825325012207
EPOCH++
Elapsed time for 100 iterations: 1.98651123046875
Elapsed time for 100 iterations: 1.0247056484222412
Elapsed time for 100 iterations: 1.1780986785888672
Elapsed time for 100 iterations: 1.7233843803405762
Elapsed time for 100 iterations: 3.0474486351013184
Elapsed time for 100 iterations: 3.7090988159179688

I tested this both with a single-epoch dataset of which I capture the tf.errors.OutOfRangeError  followed by re-initilaizing the dataset, and with a dataset.repeat().

So my question now essentially is: **what is happening inside a dataset when shuffling that makes it so much slower?** (apologies if that's in the documentation and I missed it).
Or **what am I doing wrong in its usage?**

My dataset pipeline is similar to @jrbtaylor's, I tested with the MNIST data with a batch size of 100.

> mnist_imgs = tf.constant(img_filenames) # img_filenames is a python list of strings
   mnist_labels = tf.constant(labels)
   dataset = tf.data.Dataset.from_tensor_slices((mnist_imgs, mnist_labels))
   dataset = dataset.shuffle(epoch_size) # OPTIONAL
   dataset = dataset.repeat(nb_epochs_to_do)
   dataset = dataset.map(_parse_function,num_parallel_calls=FLAGS.num_parallel_cores) # where _parse_function reads and decodes the image files
   dataset = dataset.batch(FLAGS.batch_size)
   dataset = dataset.prefetch(1)
   iterator = dataset.make_initializable_iterator()


Thank you.

Kenneth
